### DISCUSSION TITLE
Steps for loading CVRPLIB into Ecole as an Instance

### DISCUSSION THREAD

**Discussion by ndrwnaguib** [ds4dm/ecole#287] [Questions]
Hi, 

I am trying to load [CVRPLIB](http://vrp.galgos.inf.puc-rio.br/index.php/en/) into Ecole. However, I do not think designing it as an instance generator is the way. I view it as a loader instead. I'm not sure what steps I should follow to achieve that, I'd appreciate any pointers. 

Off topic: if you believe there would be considerable use for these datasets, I also would like to contribute it to Ecole problem instances.

Thank you.

**Comment by AntoinePrv** [comment]
Hi @ndrwnaguib 

It seems these files are of type `.vrp`, but I am not sure what their structure is / how to read them. Do you know how they are typically read? Do you have a reference to a parser/loader that can read these files?
I imagine that if they're not written as MILPs, it's that their formulation would be too big. Hence it's difficult to parse them into a `Model` without knowing what solution method would be used (decomposition methods most likely).

**Reply by ndrwnaguib** [reply]
> It seems these files are of type .vrp, but I am not sure what their structure is / how to read them. Do you know how they are typically read?

The structure of all VRP files included in CVRPLIB is the same, which is:

| Attribute | Description |
| ----------- | --------------- |
| Name | CVRP Instance Name. For example `A-n80-k10`, means `80` customers and `10` vehicles |
| Comment | Notes about the instance |
| Type | Problem Type (i.e., CVRP) |
| Dimension | Number of customers |
| Edge Weight Type | EUC_2D | 
| CAPACITY | Represents the maximum capacity for each vehicle. | 
| NODE_COORD_SECTION | The coordinates in the Euclidean Space |
| DEMAND_SECTION | The demand requested by each node/customer |
| DEPOT_SECTION | The location of the depot |

> Do you have a reference to a parser/loader that can read these files?

[SCIP does have a loader for VRP problems](https://www.scipopt.org/doc/html/main__vrp_8cpp_source.php):

> I imagine that if they're not written as MILPs, it's that their formulation would be too big. Hence it's difficult to parse them into a Model without knowing what solution method would be used (decomposition methods most likely).

I believe it can be solved using decomposition, after all it's an MIP problem and B&B is used to solve it as well. I would say the closest analogy in your examples is the `CapacitatedFacilityLocation`.

**Reply by AntoinePrv** [reply]
The file you are pointing to is an example file, not part of the library, so we would need to duplicate it.
The file also use a custom pricer. While we are interested in supporting this, it is not something that can be done quickly. Any help is appreciated.

**Reply by ndrwnaguib** [reply]
Great! I'd love to help. Can you please how can I? Do I prepare a PR? is there a contributing guide? :))

**Reply by AntoinePrv** [reply]
Sorry for the delay. Unfortunately, Ecole is going in a slow mode, and there is nobody that can look more in details into this.
At a first look, it seem to be that we cannot "just load" these instances like regular ones, otherwise SCIP would have added a loader for them.

**Reply by ndrwnaguib** [reply]
> Sorry for the delay. Unfortunately, Ecole is going in a slow mode, and there is nobody that can look more in details into this.

I see; I will try to work on that on my own until Ecole's development is active again. Thanks for letting me know.

> At a first look, it seem to be that we cannot "just load" these instances like regular ones, otherwise SCIP would have added a loader for them.

I'm not sure how is this different than the regular ones; I believe SCIP has a loader for CVRPLIB datasets in https://www.scipopt.org/doc/html/main__vrp_8cpp_source.php. Well, it's an example, but it should work, right?

**Reply by AntoinePrv** [reply]
My uncertainty is about the pricer on line 425. This makes me wonder that the problem loaded without the pricer might not be a complete one (not an expert here). 
So I don't know if we should load them with the pricer or leave that out.
